compiler/src/dotty/tools/dotc/core/quoted/PickledQuotes.scala

```diff
@@ -61,44 +61,117 @@ object PickledQuotes {
   }.apply(tp)
 
   /** Unpickle the tree contained in the TastyExpr */
-  def unpickleExpr(tasty: PickledQuote, args: PickledArgs)(implicit ctx: Context): Tree = {
+  def unpickleExpr(tasty: PickledQuote, splices: PickledArgs)(implicit ctx: Context): Tree = {
     val tastyBytes = TastyString.unpickle(tasty)
-    val unpickled = unpickle(tastyBytes, args, isType = false)(ctx.addMode(Mode.ReadPositions))
-    /** Force unpickling of the tree, removes the spliced type `@quotedTypeTag type` definitions and dealiases references to `@quotedTypeTag type` */
-    val forceAndCleanArtefacts = new TreeMap {
-      override def transform(tree: tpd.Tree)(implicit ctx: Context): tpd.Tree = tree match {
-        case Block(stat :: rest, expr1) if stat.symbol.hasAnnotation(defn.InternalQuoted_QuoteTypeTagAnnot) =>
-          assert(rest.forall { case tdef: TypeDef => tdef.symbol.hasAnnotation(defn.InternalQuoted_QuoteTypeTagAnnot) })
-          transform(expr1)
-        case tree => super.transform(tree).withType(dealiasTypeTags(tree.tpe))
-      }
-    }
-    forceAndCleanArtefacts.transform(unpickled)
+    val unpickled = unpickle(tastyBytes, splices, isType = false)(ctx.addMode(Mode.ReadPositions))
+    val Inlined(call, Nil, expnasion) = unpickled
+    val inlineCtx = inlineContext(call)
+    val expansion1 = spliceTypes(expnasion, splices)(using inlineCtx)
+    val expansion2 = spliceTerms(expansion1, splices)(using inlineCtx)
```

Comment:
Small opt: can we just do one pass for splicing?

Reply:
No, that was the first approach I tried. We need the types to be known when we expand the term holes as some of them may take some type parameters that the user inspects.  Note that `spliceTypes` is a no-op if the quote does not have type splices.